File: sdk-java/src/test/java/io/littlehorse/sdk/common/config/ResourceExhaustedRetryInterceptorTest.java
package io.littlehorse.sdk.common.config;

import static org.assertj.core.api.Assertions.assertThat;

import com.google.protobuf.Any;
import com.google.protobuf.util.Durations;
import com.google.rpc.Code;
import com.google.rpc.RetryInfo;
import io.grpc.Metadata;
import io.grpc.Status;
import io.grpc.protobuf.StatusProto;
import io.littlehorse.sdk.common.config.retryinterceptor.ResourceExhaustedRetryInterceptor;
import org.junit.jupiter.api.Test;

class ResourceExhaustedRetryInterceptorTest {

    @Test
    void shouldExtractRetryDelayFromStatusDetails() {
        RetryInfo retryInfo =
                RetryInfo.newBuilder().setRetryDelay(Durations.fromMillis(1500)).build();
        com.google.rpc.Status statusDetails = com.google.rpc.Status.newBuilder()
                .setCode(Code.RESOURCE_EXHAUSTED.getNumber())
                .addDetails(Any.pack(retryInfo))
                .build();
        Metadata trailers = StatusProto.toStatusRuntimeException(statusDetails).getTrailers();

        Long retryDelayMillis =
                ResourceExhaustedRetryInterceptor.getRetryDelayMillis(Status.RESOURCE_EXHAUSTED, trailers);

        assertThat(retryDelayMillis).isEqualTo(1500L);
    }

    @Test
    void shouldIgnoreStatusesWithoutRetryInfo() {
        Metadata trailers = new Metadata();

        Long retryDelayMillis =
                ResourceExhaustedRetryInterceptor.getRetryDelayMillis(Status.RESOURCE_EXHAUSTED, trailers);

        assertThat(retryDelayMillis).isNull();
    }

    @Test
    void shouldIgnoreNonQuotaStatuses() {
        RetryInfo retryInfo =
                RetryInfo.newBuilder().setRetryDelay(Durations.fromMillis(1500)).build();
        com.google.rpc.Status statusDetails = com.google.rpc.Status.newBuilder()
                .setCode(Code.RESOURCE_EXHAUSTED.getNumber())
                .addDetails(Any.pack(retryInfo))
                .build();
        Metadata trailers = StatusProto.toStatusRuntimeException(statusDetails).getTrailers();

        Long retryDelayMillis = ResourceExhaustedRetryInterceptor.getRetryDelayMillis(Status.UNAVAILABLE, trailers);

        assertThat(retryDelayMillis).isNull();
    }
}

File: sdk-dotnet/LittleHorse.Sdk.Tests/Retry/ResourceExhaustedRetryInterceptorTest.cs
using System;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Google.Rpc;
using Grpc.Core;
using Xunit;
using GrpcStatus = Grpc.Core.Status;

namespace LittleHorse.Sdk.Tests.Retry;

public class ResourceExhaustedRetryInterceptorTest
{
    [Fact]
    public void ItExtractsRetryDelayFromStatusDetails()
    {
        var retryInfo = new RetryInfo
        {
            RetryDelay = Duration.FromTimeSpan(TimeSpan.FromSeconds(2))
        };
        var statusDetails = new Google.Rpc.Status
        {
            Code = (int)Code.ResourceExhausted
        };
        statusDetails.Details.Add(Any.Pack(retryInfo));

        var rpcException = new RpcException(
            new GrpcStatus(StatusCode.ResourceExhausted, "quota exceeded"),
            new Metadata { { ResourceExhaustedRetryInterceptor.StatusDetailsKey, statusDetails.ToByteArray() } });

        var didExtract = ResourceExhaustedRetryInterceptor.TryGetRetryDelay(rpcException, out var retryDelay);

        Assert.True(didExtract);
        Assert.Equal(TimeSpan.FromSeconds(2), retryDelay);
    }

    [Fact]
    public void ItIgnoresResourceExhaustedWithoutRetryInfo()
    {
        var rpcException = new RpcException(
            new GrpcStatus(StatusCode.ResourceExhausted, "quota exceeded"),
            new Metadata());

        var didExtract = ResourceExhaustedRetryInterceptor.TryGetRetryDelay(rpcException, out _);

        Assert.False(didExtract);
    }
}
